src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionInvoker.cs

```diff
@@ -709,125 +1251,177 @@ private Task InvokeAllExceptionFiltersAsync()
             }
 
-            await InvokeResultFilterAsync();
+            await InvokeNextResultFilterAsync();
 
             Debug.Assert(_resultExecutedContext != null);
             return _resultExecutedContext;
         }
 
-        private async Task InvokeResultFilterAsync()
+        private async Task InvokeResultAsync(IActionResult result)
         {
-            Debug.Assert(_resultExecutingContext != null);
+            var controllerContext = _controllerContext;
+
+            _diagnosticSource.BeforeActionResult(controllerContext, result);
 
             try
             {
-                var item = _cursor.GetNextFilter<IResultFilter, IAsyncResultFilter>();
-                if (item.FilterAsync != null)
-                {
-                    _diagnosticSource.BeforeOnResultExecution(_resultExecutingContext, item.FilterAsync);
-
-                    await item.FilterAsync.OnResultExecutionAsync(_resultExecutingContext, InvokeResultFilterAwaitedAsync);
-
-                    if (_resultExecutedContext == null || _resultExecutingContext.Cancel == true)
-                    {
-                        // Short-circuited by not calling next || Short-circuited by setting Cancel == true
-                        _logger.ResourceFilterShortCircuited(item.FilterAsync);
-
-                        _resultExecutedContext = new ResultExecutedContext(
-                            _resultExecutingContext,
-                            _filters,
-                            _resultExecutingContext.Result,
-                            _controller)
-                        {
-                            Canceled = true,
-                        };
-                    }
-
-                    _diagnosticSource.AfterOnResultExecution(_resultExecutedContext, item.FilterAsync);
-                }
-                else if (item.Filter != null)
-                {
-                    _diagnosticSource.BeforeOnResultExecuting(_resultExecutingContext, item.Filter);
+                await result.ExecuteResultAsync(controllerContext);
+            }
+            finally
+            {
+                _diagnosticSource.AfterActionResult(controllerContext, result);
+            }
+        }
 
-                    item.Filter.OnResultExecuting(_resultExecutingContext);
+        private static void Rethrow(ResourceExecutedContext context)
+        {
+            if (context == null)
+            {
+                return;
+            }
 
-                    _diagnosticSource.AfterOnResultExecuting(_resultExecutingContext, item.Filter);
+            if (context.ExceptionHandled)
+            {
+                return;
+            }
 
-                    if (_resultExecutingContext.Cancel == true)
-                    {
-                        // Short-circuited by setting Cancel == true
-                        _logger.ResourceFilterShortCircuited(item.Filter);
+            if (context.ExceptionDispatchInfo != null)
```

Comment:
We only check for context.Exception in many of the states, is there a condition that when Exception != null ExceptionDispatchInfo != null too? If not, should we do a different check on the places where we check for Exception? My understanding is that ExceptionDispatchInfo is for async cases where Exception is for Sync cases.